models/surveys.go

```diff
@@ -29,53 +126,177 @@ func AllSurveys() ([]*SurveySummary, error) {
+func (api *API) GetSurveyByShortName(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["shortName"]
+	surveyRow := api.GetSurveyByShortNameStmt.QueryRow(id)
 	survey := new(Survey)
-	err := db.QueryRow("SELECT id, shortname, longname, surveyref from survey.survey WHERE LOWER(shortName) = LOWER($1)", shortName).Scan(&survey.ID, &survey.ShortName, &survey.LongName, &survey.Reference)
+	err := surveyRow.Scan(&survey.ID, &survey.ShortName, &survey.LongName, &survey.Reference)
 	if err != nil {
-		return nil, err
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "get survey by shortname query failed", http.StatusInternalServerError)
+		}
 	}
 
-	return survey, nil
+	data, _ := json.Marshal(survey)
+	w.Write(data)
+
 }
 
-// GetSurveyByReference returns the details of the survey identified by the string reference.
-func GetSurveyByReference(reference string) (*Survey, error) {
+// GetSurveyByReference returns the details of the survey identified by the string ref.
+func (api *API) GetSurveyByReference(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["ref"]
+	surveyRow := api.GetSurveyByReferenceStmt.QueryRow(id)
 	survey := new(Survey)
-	err := db.QueryRow("SELECT id, shortname, longname, surveyref from survey.survey WHERE LOWER(surveyref) = LOWER($1)", reference).Scan(&survey.ID, &survey.ShortName, &survey.LongName, &survey.Reference)
+	err := surveyRow.Scan(&survey.ID, &survey.ShortName, &survey.LongName, &survey.Reference)
 	if err != nil {
-		return nil, err
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "get survey by reference query failed", http.StatusInternalServerError)
+		}
+	}
+
+	data, _ := json.Marshal(survey)
+	w.Write(data)
+
+}
+
+// AllClassifierTypeSelectors returns all the classifier type selectors for the survey identified by the string surveyID. The classifier type selectors are returned in ascending order.
+func (api *API) AllClassifierTypeSelectors(w http.ResponseWriter, r *http.Request) {
+	// We need to run a query first to check if the survey exists so an HTTP 404 can be correctly
+	// returned if it doesn't exist. Without this check an HTTP 204 is incorrectly returned for an
+	// invalid survey ID.
+	vars := mux.Vars(r)
+	surveyId := vars["surveyId"]
+
+	err := api.getSurveyID(surveyId)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			http.Error(w, "Survey not found", http.StatusNotFound)
+		} else {
+			http.Error(w, "Error getting list of classifier type selectors for survey '"+surveyId+"' - "+err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	// Now we can get the classifier type selector records.
+	rows, err := api.GetClassifierTypeSelectorStmt.Query(surveyId)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			http.Error(w, "Classifier type selector not found", http.StatusNotFound)
```

Comment:
The previous version of the survey service had a "RESTError" model to represent a 404 not found - i think other services calling expect this data structure so we need to use it for the time being